keps/349-avoid-coscheduling/README.md

```diff
@@ -0,0 +1,350 @@
+- guarantee that two workloads would not schedule pods concurrently. Example
+scenarios in which two workloads may still concurrently schedule their pods:
+  - when succeeded pods are replaced with new because job's parallelism is less than its completions;
+  - when a failed pod gets replaced.
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+We introduce a mechanism to mark a Job as blocking admission of further
+workloads until the first batch of its pods is scheduled, more precisely, until:
+`job.status.ready + job.status.succeeded >= job.spec.parallelism`.
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+As a machine learning researcher I want to run a large workload, comprising
+thousands of pods which need to all run at the same time. Scheduling all pods
+in this workload takes up to 20min and I want to have a guarantee that other
+workloads scheduled within this time interval would not delay execution of my
+workload.
+
+My use can be supported by adding `kueue.x-k8s.io/wait-for-pod-scheduling`
+annotation to the jobs I run.
+
+#### Story 2
+
+As an administrator of the batch processing platform I want to be able to
+configure kueue quotas for on-demand resources a little higher than what is
+available in the quiet period. However, I don't want to risk two workloads
+deadlock in case of issues with provisioning nodes to satisfy the configured
+quota.
+
+My use can be supported by adding `kueue.x-k8s.io/wait-for-pod-scheduling`
+annotation to all jobs. I could also designate a single local queue
+`wait-for-pod-scheduling-queue` to be responsible for such workloads and only
+add the annotation based on the value of the `kueue.x-k8s.io/queue-name`
+annotation. This way we could allow for coscheduling of workloads if the local
+queue is not used.
+
+### Notes/Contraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+If the workload fails to schedule its pods it could block admission of other
+workloads indefinitely.
+
+In order to mitigate this risk we terminate workloads which don't have enough
```

Comment:
which timeout or where is it defined? And what happens to the workload? is it put back into the queue?

Reply:
The timeout would be defined at a Job level by annotation. I see two options:
- terminate the job (can be now done using the TargetFailure condition) and delete the workload, or
- suspend the job and requeue the workload. 

However, the lack of timeout is a problem also currently as accumulated admitted workloads which fail to schedule will block further admission. Thus, I'm thinking of deferring the work and design properly under a dedicated KEP.